# SQLQuest

Catherine Devlin

Ohio LinuxFest 2015

https://github.com/catherinedevlin/sql_quest

![knights](img/knights.jpg)


# Adventure Synopsis

* Basic: CSV
* Advanced: JSON
* Expert: Relational

![duel with monster](img/monster_duel.jpg)

# In the beginning...

![D&D character sheet](img/charsheet.png)

# Into Electronica

All but useless ectronically

In [1]:
!libreoffice data/aelfryth.odt

# CSV

In [7]:
!libreoffice data/party.ods
!cat data/party.csv





Wigstan,Thief,Runaway thrall,3,14,12,9,15,14,10,cudgel,1,3,0.1,,wood,,cookpot,1,10,0.3,,iron,,dagger,1,1,2,,iron,,torches,3,2,0.02,,wood,

In [8]:
from csv import DictReader
from pprint import pprint
with open('data/party.csv') as infile:
    party = list(DictReader(infile))
    
pprint(party)

[{'charisma': '14',
  'class': 'Fighter',
  'constitution': '15',
  'dexterity': '15',
  'equip 1 cost': '7',
  'equip 1 magic': '',
  'equip 1 materials': 'wood, cloth',
  'equip 1 name': 'shield',
  'equip 1 notes': '',
  'equip 1 quantity': '1',
  'equip 1 weight each': '10',
  'equip 2 cost': '4',
  'equip 2 magic': '',
  'equip 2 materials': 'cloth',
  'equip 2 name': 'Tent',
  'equip 2 notes': 'sleeps 4',
  'equip 2 quantity': '1',
  'equip 2 weight each': '15',
  'equip 3 cost': '30',
  'equip 3 magic': '',
  'equip 3 materials': 'wood',
  'equip 3 name': 'bow',
  'equip 3 notes': '',
  'equip 3 quantity': '1',
  'equip 3 weight each': '2',
  'equip 4 cost': '0.1',
  'equip 4 magic': '',
  'equip 4 materials': 'wood, iron',
  'equip 4 name': 'arrows',
  'equip 4 notes': '',
  'equip 4 quantity': '12',
  'equip 4 weight each': '0.1',
  'intelligence': '11',
  'level': '5',
  'name': 'Aelfryth',
  'role': 'Thane',
  'strength': '16',
  'wisdom': '14'},
 {'charisma': '12',
  'class

# Answering questions

How much weight are the party's fighters carrying?

In [3]:
total_weight = 0
for character in party:
    if character['class'] == 'Fighter':
        for eq_index in range(1, 4):
            try:
                quantity = int(character['equip %d quantity' % eq_index])
                weight_each = float(character['equip %d weight each' % eq_index])
                total_weight += quantity * weight_each
            except (TypeError, ValueError):
                pass
print(total_weight)

56.0


# Save vs. Confusion

* custom program
* hardcoded number of equipment slots

![lost](img/lost.jpg)

# Relational (from CSV)

In [10]:
!libreoffice data/party.multisheet.ods

In [11]:
from csv import DictReader
from pprint import pprint
with open('data/party.multisheet/stats-Table 1.csv') as infile:
    stats = list(DictReader(infile))
with open('data/party.multisheet/equipment-Table 1.csv') as infile:
    equipment = list(DictReader(infile))
pprint(stats)

[{'charisma': '14',
  'class': 'Fighter',
  'constitution': '15',
  'dexterity': '15',
  'intelligence': '11',
  'level': '5',
  'name': 'Aelfryth',
  'role': 'Thane',
  'strength': '16',
  'wisdom': '14'},
 {'charisma': '12',
  'class': 'Fighter',
  'constitution': '16',
  'dexterity': '11',
  'intelligence': '9',
  'level': '4',
  'name': 'Godric',
  'role': 'Warband Member',
  'strength': '17',
  'wisdom': '7'},
 {'charisma': '11',
  'class': 'Sourceror',
  'constitution': '12',
  'dexterity': '14',
  'intelligence': '17',
  'level': '5',
  'name': 'Leofflaed',
  'role': 'Enigma',
  'strength': '8',
  'wisdom': '11'},
 {'charisma': '10',
  'class': 'Thief',
  'constitution': '14',
  'dexterity': '15',
  'intelligence': '12',
  'level': '3',
  'name': 'Wigstan',
  'role': 'Runaway Thrall',
  'strength': '14',
  'wisdom': '9'}]


In [24]:
total_weight = 0
for character in stats:
    if character['class'] == 'Fighter':
        for itm in equipment:
            if itm['owner'] == character['name']:
                try:
                    quantity = int(itm['quantity'])
                    weight_each = float(itm['weight each'])
                    total_weight += quantity * weight_each
                except (TypeError, ValueError):
                    pass
print(total_weight)

57.2


# Still ouch

![pained face](img/ouch.jpg)

# JSON

arbitrary structure

In [12]:
!cat data/party.json

[
  {
    "class": "Thief",
    "role": "Runaway Thrall",
    "level": 3,
    "equipment": {
      "torches": {
        "quantity": 3,
        "materials": [
          "wood"
        ],
        "weight each": 2.0,
        "magic": "",
        "notes": "",
        "cost": 2.0
      },
      "cudgel": {
        "quantity": 1,
        "materials": [
          "wood"
        ],
        "weight each": 3.0,
        "magic": "",
        "notes": "",
        "cost": 3.0,
        "damage": "d6"
      },
      "cookpot": {
        "quantity": 1,
        "materials": [
          "iron"
        ],
        "weight each": 10.0,
        "magic": "",
        "notes": "",
        "cost": 10.0
      },
      "dagger": {
        "quantity": 1,
        "materials": [
          "iron"
        ],
        "weight each": 1.0,
        "magic": "",
        "notes": "",
        "cost": 1.0,
        "damage": "d4"
      }
    },
    "name": "Wigstan",
    "stats": {
      "dexterity": 15,
      "charisma": 10,
  

How much does all the fighters' equipment weigh?

In [2]:
import json
import glob
total_weight = 0
with open('data/party.json') as infile:
    for character in json.load(infile):
        if character['class'] == 'Fighter':
            for itm in character['equipment'].values():
                total_weight += (itm['quantity'] * itm['weight each'])
print(total_weight)

57.2


# Consistency

In [45]:
party.append( {'name': 'Mickey',
               'class': 'Druid/Gunslinger/Paladin/Illusionist/Assassin',
               'level': 87,
               'psionic level': 19} )

Schema enforcement: kwalify, marshmallow, ...

# Scaling

* Multiple access

## Document databases

* Mongo
* 

In [28]:
!echo "db.party.drop()" | mongo

MongoDB shell version: 2.6.3
connecting to: test
true
bye


In [29]:
!mongoimport --collection party --jsonArray party.json

connected to: 127.0.0.1
2015-09-25T19:23:17.661-0400 imported 4 objects


In [30]:
!echo "db.party.find()" | mongo

MongoDB shell version: 2.6.3
connecting to: test
{ "_id" : ObjectId("5605d765636c85fac5768a6f"), "class" : "Thief", "role" : "Runaway Thrall", "level" : 3, "equipment" : { "torches" : { "quantity" : 3, "materials" : [ "wood" ], "weight each" : 2, "magic" : "", "notes" : "", "cost" : 2 }, "cudgel" : { "quantity" : 1, "materials" : [ "wood" ], "weight each" : 3, "magic" : "", "notes" : "", "cost" : 3, "damage" : "d6" }, "cookpot" : { "quantity" : 1, "materials" : [ "iron" ], "weight each" : 10, "magic" : "", "notes" : "", "cost" : 10 }, "dagger" : { "quantity" : 1, "materials" : [ "iron" ], "weight each" : 1, "magic" : "", "notes" : "", "cost" : 1, "damage" : "d4" } }, "name" : "Wigstan", "stats" : { "dexterity" : 15, "charisma" : 10, "strength" : 14, "wisdom" : 9, "intelligence" : 12, "constitution" : 14 } }
{ "_id" : ObjectId("5605d765636c85fac5768a70"), "class" : "Fighter", "role" : "Thane", "level" : 5, "equipment" : { "bow" : { "quantity" : 1, "materials" : [ "wood" ], "weight each"

How much weight are the party's fighters carrying?

In [31]:
!echo "db.party.find({class: 'Fighter'}, {'equipment': 1, _id: 0})" | mongo

MongoDB shell version: 2.6.3
connecting to: test
{ "equipment" : { "bow" : { "quantity" : 1, "materials" : [ "wood" ], "weight each" : 2, "magic" : "", "notes" : "", "cost" : 2, "damage" : "d6" }, "Tent" : { "quantity" : 1, "materials" : [ "cloth" ], "weight each" : 15, "magic" : "", "notes" : "sleeps 4", "cost" : 15 }, "shield" : { "quantity" : 1, "materials" : [ "wood", "cloth" ], "weight each" : 10, "magic" : "", "notes" : "", "cost" : 10 }, "arrows" : { "quantity" : 12, "materials" : [ "wood", "iron" ], "weight each" : 0.1, "magic" : "", "notes" : "", "cost" : 0.1 } } }
{ "equipment" : { "torches" : { "quantity" : 6, "materials" : [ "wood" ], "weight each" : 2, "magic" : "", "notes" : "", "cost" : 2 }, "30’ rope" : { "quantity" : 1, "materials" : [ "hemp" ], "weight each" : 10, "magic" : "", "notes" : "", "cost" : 10 }, "battleaxe" : { "quantity" : 1, "materials" : [ "wood", "iron" ], "weight each" : 7, "magic" : "+1", "notes" : "", "cost" : 7, "damage" : "d8" } } }
bye


In [32]:
TODO: finish MongoDB query

SyntaxError: invalid syntax (<ipython-input-32-ddb456e8d93c>, line 1)

# RDBMS

Relational DataBase Management System

* PostgreSQL
* SQLite

# RDBMS provides

* scale
* data consistency
* query language: SQL
* tools


# DDL

Data Definition Language

![AD&D Players' Handbook](https://upload.wikimedia.org/wikipedia/en/0/09/PlayersHandbook8Cover.jpg)

Create a database and connect to it

    $ psql -h 127.0.0.1 template1 dungeonmaster
    Password for user dungeonmaster: 
    psql (9.4.1)
    Type "help" for help.

    template1=# create database dnd;
    CREATE DATABASE
    template1=# \c dnd
    You are now connected to database "dnd" as user "dungeonmaster".


In [33]:
!cat sql/ddl/create_char_tbl.sql

  CREATE TABLE character (
    name TEXT,
    class TEXT,
    role TEXT,
    level INTEGER,
    strength INTEGER,
    intelligence INTEGER,
    wisdom INTEGER,
    dexterity INTEGER,
    constitution INTEGER,
    charisma INTEGER );


    dnd=# \i sql/ddl/create_char_tbl.sql 
    CREATE TABLE

# DML

Data Manipulation Language

In [34]:
!cat sql/dml/ins_single_char.sql

INSERT INTO character (
  name, class, role, level,
  strength, intelligence, wisdom, dexterity,
  constitution, charisma)
VALUES (
  'Aelfryth', 'Fighter', 'Thane', 5,
  16, 11, 14, 15,
  15, 14);


    dnd=# \i sql/dml/ins_single_char.sql
    INSERT 0 1

# SELECT

    dnd=# SELECT * FROM character;
       name   |  class  | role  | level | strength | intelligence | wisdom | dexterity | constitution | charisma 
    ----------+---------+-------+-------+----------+--------------+--------+-----------+--------------+----------
     Aelfryth | Fighter | Thane |     5 |       16 |           11 |     14 |        15 |           15 |       14
    (1 row)

# Mass insert

    dnd=# DELETE FROM character;
    DELETE 4
    dnd=# \copy character FROM 'party.multisheet/stats-Table 1.csv' WITH csv HEADER;
    COPY 4

# Selective SELECT

    dnd=# SELECT name, class, role FROM character;
       name    |   class   |      role      
    -----------+-----------+----------------
     Aelfryth  | Fighter   | Thane
     Godric    | Fighter   | Warband Member
     Leofflaed | Sourceror | Enigma
     Wigstan   | Thief     | Runaway Thrall
    (4 rows)

# WHERE

    dnd=# SELECT name, class, role 
    dnd-# FROM   character
    dnd-# WHERE  class = 'Fighter';
       name   |  class  |      role      
    ----------+---------+----------------
     Aelfryth | Fighter | Thane
     Godric   | Fighter | Warband Member
    (2 rows)

# AND

    dnd=# SELECT name, class, role, intelligence
    dnd-# FROM   character
    dnd-# WHERE  class = 'Fighter'
    dnd-# AND    intelligence > 10;
       name   |  class  | role  | intelligence 
    ----------+---------+-------+--------------
     Aelfryth | Fighter | Thane |           11
    (1 row)

# Level up

    dnd=# SELECT name, level FROM character;
       name    | level 
    -----------+-------
     Aelfryth  |     5
     Godric    |     4
     Leofflaed |     5
     Wigstan   |     3
    (4 rows)



# UPDATE

    dnd=# UPDATE character
    dnd-# SET    level = 4
    dnd-# WHERE  name = 'Wigstan';
    UPDATE 1

    dnd=# SELECT name, level FROM character;                                                                               
     name      | level 
    -----------+-------
     Aelfryth  |     5
     Godric    |     4
     Leofflaed |     5
     Wigstan   |     4
    (4 rows)


# Unique identifiers

In [ ]:
!cat sql/dml/wigmund.sql

    dnd=# SELECT name, class, level FROM character WHERE class = 'Thief';
      name   | class | level 
    ---------+-------+-------
     Wigstan | Thief |     4
     Wigmund | Thief |     4
    (2 rows)


# Oops

    dnd=# UPDATE character SET level = level + 1 WHERE class = 'Thief';
    UPDATE 2
    dnd=# SELECT name, class, level FROM character WHERE class = 'Thief';
      name   | class | level 
    ---------+-------+-------
     Wigstan | Thief |     5
     Wigmund | Thief |     5
    (2 rows)

# Primary Key

## numeric

    dnd=# ALTER TABLE character ADD
    dnd-#   id SERIAL PRIMARY KEY;

    dnd=# SELECT id, name, class, role FROM character;
     id |   name    |   class   |      role      
    ----+-----------+-----------+----------------
      1 | Aelfryth  | Fighter   | Thane
      2 | Godric    | Fighter   | Warband Member
      3 | Leofflaed | Sourceror | Enigma
      4 | Wigstan   | Thief     | Runaway Thrall
      5 | Wigmund   | Thief     | Runaway Thrall

    dnd=# ALTER TABLE character DROP id;
    ALTER TABLE

## Text PK
  
    dnd=# ALTER TABLE character
    dnd-# ADD PRIMARY KEY (name);
    ALTER TABLE

In [46]:
!cat sql/dml/godric2.sql

INSERT INTO character (
  name, class, role, level,
  strength, intelligence, wisdom, dexterity,
  constitution, charisma)
VALUES (
  'Godric', 'Ranger', 'Guide', 4,
  14, 15, 13, 15,
  16, 7);


    dnd=# \i sql/dml/godric2.sql 
    psql:sql/dml/godric2.sql:8: ERROR:  duplicate key value violates unique constraint "character_pkey"
    DETAIL:  Key (name)=(Godric) already exists.

# Relations



In [ ]:
!cat sql/ddl/create_equip_tbl.sql

    dnd=# \i sql/ddl/create_equip_tbl.sql 
    CREATE TABLE
    dnd=# \copy equipment FROM 'party.multisheet/equipment-Table 1.csv' WITH csv HEADER
    COPY 14

# Join

All Leofflaed's gear

    dnd=# SELECT c.name, e.name, e.magic
    FROM   character c
    JOIN   equipment e ON (e.owner = c.name)
    WHERE  c.name = 'Leofflaed';
    
       name    |       name        | magic 
    -----------+-------------------+-------
     Leofflaed | Oil of Revelation | Y
     Leofflaed | Unfailing Flint   | Y
     Leofflaed | lamp              | 

Weight of all fighters' gear

    dnd=# SELECT SUM(quantity * weight_each)
    dnd-# FROM   equipment e
    dnd-# JOIN   character c ON (e.owner = c.name)
    dnd-# WHERE  class = 'Fighter';
     sum  
    ------
     57.2
    (1 row)

# Constraints

![Chess players](img/chess.jpg)

In [37]:
!cat sql/dml/mickey.sql

INSERT INTO character (
  name, class, role, level,
  psionic_level,
  strength, intelligence, wisdom, dexterity,
  constitution, charisma)
VALUES (
  'Mickey', 'Druid/Gunslinger/Paladin/Illusionist/Assassin', 'Demigod', 87,
  19,
  19, 19, 18, 18,
  19, 17);



    ERROR:  column "psionic_level" of relation "character" does not exist
    LINE 3:   psionic_level,

# Column constraints

      CREATE TABLE character (
        name TEXT,
        class VARCHAR(14),
        ...

    dnd=# ALTER TABLE character ALTER COLUMN class TYPE VARCHAR(14);
    
    INSERT INTO character (
    ...

    ERROR:  value too long for type character varying(14)

# Foreign key constraints

# Check constraints

      CREATE TABLE character (
        ...
        level INTEGER,
        ...
        
      dnd=# ALTER TABLE character ADD CONSTRAINT reaonable_level CHECK (level < 15);

    ERROR:  new row for relation "character" violates check constraint "character_level_check"
    DETAIL:  Failing row contains (Mickey, Druid, Demigod, 87, 19, 19, 18, 18, 19, 17).        

# Transactions

![transaction](img/transaction.jpg)

In [42]:
!cat sql/dml/thief_inventories.sql

SELECT name, owner
FROM   equipment
WHERE  owner LIKE 'Wig%';


In [ ]:
dnd=# \i sql/dml/thief_inventories.sql
  name   |  owner  
---------+---------
 cudgel  | Wigstan
 cookpot | Wigstan
 dagger  | Wigstan
 torches | Wigstan
(4 rows)


    dnd=# UPDATE equipment
    dnd-# SET    owner = 'Wigmund'
    dnd-# WHERE  owner = 'Wigstan'
    dnd-# AND    name = 'cudgel';
    UPDATE 1
    dnd=# \i sql/dml/thief_inventories.sql 
      name   |  owner  
    ---------+---------
     cookpot | Wigstan
     dagger  | Wigstan
     torches | Wigstan
     cudgel  | Wigmund
    (4 rows)

    dnd=# BEGIN TRANSACTION;
    BEGIN
    dnd=# UPDATE equipment
    dnd-# SET    owner = 'Wigmund'
    dnd-# WHERE  owner = 'Wigstan'
    dnd-# AND    name = 'dagger';
    UPDATE 1
    dnd=# SELECT name, owner
    dnd-# FROM   equipment
    dnd-# WHERE  owner LIKE 'Wig%';
      name   |  owner  
    ---------+---------
     cookpot | Wigstan
     torches | Wigstan
     cudgel  | Wigmund
     dagger  | Wigmund
    (4 rows)


    dnd=# ROLLBACK;
    ROLLBACK
    dnd=# SELECT name, owner
    dnd-# FROM   equipment
    dnd-# WHERE  owner LIKE 'Wig%';
      name   |  owner  
    ---------+---------
     cookpot | Wigstan
     torches | Wigstan
     dagger  | Wigstan
     cudgel  | Wigmund
    (4 rows)
    
opposite is `COMMIT`

    dnd=# update equipment set owner = 'Wigstan' WHERE owner = 'Wigmund';
    UPDATE 0
    dnd=# 

# Aggregate functions

    dnd=# SELECT MAX(charisma) FROM character;
     max 
    -----
      14
    (1 row)
    
Think line count

    dnd=# SELECT name, MAX(charisma) FROM character;
    ERROR:  column "character.name" must appear in the GROUP BY clause or be used in an aggregate function
    LINE 1: SELECT name, MAX(charisma) FROM character;

## Subquery

    dnd=# SELECT name, charisma
    dnd-# FROM   character
    dnd-# WHERE  charisma = (SELECT MAX(charisma) FROM character);
       name   | charisma 
    ----------+----------
     Aelfryth |       14
    (1 row)


## GROUP BY

    dnd=# SELECT owner, SUM(cost) 
    dnd-# FROM   equipment
    dnd-# GROUP BY owner
    dnd-# ORDER BY SUM(cost) DESC;
       owner   |  sum   
    -----------+--------
     Leofflaed | 550.50
     Aelfryth  |  41.10
     Godric    |   5.12
     Wigstan   |   2.42
    (4 rows)


## Outer joins

    dnd=# SELECT c.name, c.class, SUM(e.cost)
    dnd-# FROM   character c
    dnd-# JOIN   equipment e ON (e.owner = c.name)
    dnd-# GROUP BY c.name, c.class
    dnd-# ORDER BY SUM(e.cost) DESC;
       name    |   class   |  sum   
    -----------+-----------+--------
     Leofflaed | Sourceror | 550.50
     Aelfryth  | Fighter   |  41.10
     Godric    | Fighter   |   5.12
     Wigstan   | Thief     |   2.42
    (4 rows)


    dnd=# SELECT c.name, c.class, SUM(e.cost)
    FROM   character c
    LEFT OUTER JOIN equipment e ON (e.owner = c.name)
    GROUP BY c.name, c.class
    ORDER BY SUM(e.cost) DESC;
       name    |   class   |  sum   
    -----------+-----------+--------
     Wigmund   | Thief     |       
     Leofflaed | Sourceror | 550.50
     Aelfryth  | Fighter   |  41.10
     Godric    | Fighter   |   5.12
     Wigstan   | Thief     |   2.42
    (5 rows)
    
... optionally, `NULLS LAST`


# Traps

![trap book](img/grimtooth.jpg)

# The NULL trap (3-value logic)


    dnd=# SELECT name, magic 
    dnd-# FROM   equipment
    dnd-# WHERE  magic IS NOT NULL;
           name        | magic 
    -------------------+-------
     battleaxe         | +1
     Oil of Revelation | Y
     Unfailing Flint   | Y
    (3 rows)

    dnd=# SELECT name, magic
    dnd-# FROM   equipment
    dnd-# WHERE  magic = NULL;
     name | magic 
    ------+-------
    (0 rows)


    dnd=# SELECT 1 = 1;
     ?column? 
    ----------
     t
    (1 row)

    dnd=# SELECT 1 = 2;
     ?column? 
    ----------
     f
    (1 row)

    dnd=# SELECT NULL = NULL;
     ?column? 
    ----------

    (1 row)

    dnd=# SELECT NULL != NULL;
     ?column? 
    ----------

    (1 row)

    dnd=# SELECT c.name
    dnd-# FROM   character c
    dnd-# JOIN   equipment e ON (e.owner = c.name)
    dnd-# AND    e.magic != NULL;
     name 
    ------
    (0 rows)
    
    dnd=# SELECT c.name AS owner, e.name                                                           dnd=# FROM   character c                                                                       dnd=# JOIN   equipment e ON (c.name = e.owner)                                                 dnd=# WHERE  magic IS NULL;
       owner   |   name   
    -----------+----------
     Aelfryth  | shield
     Aelfryth  | Tent
     Aelfryth  | bow
     Aelfryth  | arrows
     Godric    | 30’ rope
     Godric    | torches
     Leofflaed | lamp
     Wigstan   | cookpot
     Wigstan   | torches
     Wigstan   | dagger
     Wigmund   | cudgel
    (11 rows)


## Missing WHERE trap

    dnd=# SELECT name, level
    dnd-# FROM   character;
       name    | level 
    -----------+-------
     Aelfryth  |     5
     Godric    |     4
     Leofflaed |     5
     Wigstan   |     5
     Wigmund   |     5
    (5 rows)

    dnd=# UPDATE character
    dnd-# SET    level = 6;
    UPDATE 5
    dnd=# SELECT name, level
    dnd-# FROM   character;
       name    | level 
    -----------+-------
     Aelfryth  |     6
     Godric    |     6
     Leofflaed |     6
     Wigstan   |     6
     Wigmund   |     6
    (5 rows)


## Unqualified `DELETE` = apocalypse

    dnd=# BEGIN TRANSACTION;
    BEGIN
    dnd=# DELETE FROM equipment;
    DELETE 14
    dnd=# SELECT * FROM equipment;
     name | quantity | weight_each | cost | magic | materials | notes | owner | damage 
    ------+----------+-------------+------+-------+-----------+-------+-------+--------
    (0 rows)

    dnd=# ROLLBACK;
    ROLLBACK

# Tools

* [GUIs](https://wiki.postgresql.org/wiki/Community_Guide_to_PostgreSQL_GUI_Tools)

* [Web interfaces](http://www.postgresqlstudio.org/)

* Automatic APIs: [HTSQL](http://htsql.org/), [Postgrest](https://github.com/begriffs/postgrest)...

# Triggers

![woman with crossbow](img/crossbow2.jpg)

# Take 500 XP

Questions?

catherinedevlin.blogspot.com
[github.com/catherinedevlin/sql_quest](https://github.com/catherinedevlin/sql_quest)

Images (except book covers) from [British Library's public release](https://www.flickr.com/photos/britishlibrary)